Tag tree actions

will-moore opened the pull request:
This PR updates the web with the same Tag & Tag set functionality as Insight.

![screen shot 2016-03-08 at 16 33 26](https://cloud.githubusercontent.com/assets/900055/13608188/939b00fe-e54b-11e5-9fd8-062062a1ac67.png)

To test:
 - Browse tags tree
 - Creation of tags / tag sets should appear as orphans in tree, unless tag is created with tag set selected (should be created as child of tag set).
     - Create Tag and Tag set from toolbar buttons
     - Create Tag and Tag set from right-click menu
 - Cut, Copy and Paste all data types, from toolbar and Right-click menu:
    - Tag / Tagset
    - Dataset / Project
    - Image / Dataset
    - Plate / Screen
 - Drag and Drop all data types that can be Cut as above
 - Tagged objects can also be Copied from below Tag (NOT Cut, as in Insight)
 - Delete Tag & Tag set

NB: @pwalczysko, @bramalingam and I had a discussion about whether you should be able to Cut OR Copy orphaned objects E.g.:
 - Dataset in P/D/I tree
 - Orphaned Image in P/D/I tree
 - Tag in Tag tree

Generally agreed that you should be able to Copy these (not current policy in Insight or Web & not changed in this PR).
Re: this PR, we allow to Copy:
 - Tagged Project/Dataset/Image etc in Tag tree (where you are not Cutting the parent link)
This is conceptually similar to Copy of an Orphaned object (also not Cutting any parent link).

will-moore commented:
@pwalczysko Could you check whether https://trac.openmicroscopy.org/ome/ticket/13020 is fixed in this PR? I think it is. Thanks.

pwalczysko commented:
@will-moore : Just checked, yes, the https://trac.openmicroscopy.org/ome/ticket/13020 works as expected for all ro, ra, rw groups and for any type of users.

pwalczysko commented:
Creation of new tags and Tagsets - works as expected for icons and right-click in Tags Tab.

pwalczysko commented:
Suggestion: Make the Tagset icon more distinctive from Tag icon (atm they look very similar, hard to see on first look which is which).
![screen shot 2016-03-11 at 11 01 25](https://cloud.githubusercontent.com/assets/2478303/13700394/a29e6332-e778-11e5-83f1-281ae5cc49a6.png)

pwalczysko commented:
Possible RFE see https://trac.openmicroscopy.org/ome/ticket/13172#ticket (not for this PR).

pwalczysko commented:
>Cut, Copy and Paste all data types, from toolbar and Right-click menu:

>    Tag / Tagset
>    Dataset / Project
>    Image / Dataset
>    Plate / Screen

> Drag and Drop all data types that can be Cut as above

These workflows worked as expected on your own data.

pwalczysko commented:
> Tagged objects can also be Copied from below Tag (NOT Cut, as in Insight)

After discussion with @will-moore , this means that if you have an image, plate, or dataset under a Tag, you can copy them to a dataset, screen or project (respectively) in the Tag Tab. Insight shows at the same place only Cut in the context menu. As the procedure behaves clearly as Copy, I think the behaviour in web is right.

pwalczysko commented:
Bug: discrepancy between the right-click menu and icons:
   - have an orphaned tag in Tags Tab
   - right click on it -> you have only ``Cut link`` enabled
   - in the icons above the left hand pane you have also ``Copy`` enabled
   - when using the icon, it ``copies`` the link, leaving the orphaned Tag in its original place as well as putting it into the selected Tagset, but after refresh, the orphaned Tag vanishes
   - the same workflow in Insight has just Copy enabled, but, unfortunately, it does not actually work (warning ``the nodes to copy cannot be added to the selected nodes``)

Recommend: do not follow insight in this particular workflow, as it is both illogical and dysfunctional, but SYNCHRONIZE the icons and right-click menu.

gusferguson commented:
@will-moore 

Tested using https://cowfish.openmicroscopy.org/merge/webclient/usertags/ user-3 read-only-1 Mac Safari

- Browse tags tree

Behaves as expected

- Creation of tags / tag sets should appear as orphans in tree, unless tag is created with tag set selected (should be created as child of tag set).
    - Create Tag and Tag set from toolbar buttons
    - Create Tag and Tag set from right-click menu

Functionality behaves as described.
Minor point: if a tag within a tagset is selected and Create Tag performed I expected the new tag to be created in the same tagset, not orphaned as it does at present. Not a major issue, but I think this would be expected behaviour e.g. from behaviour in Finder/Explorer where new object is created in current location in the hierarchy.
Noted that current behaviour is consistent with Project-Dataset behaviour so same applies to this.

- Cut, Copy and Paste all data types, from toolbar and Right-click menu:

-- Tag / Tagset

All work except when cutting link with tag in tagset - the cut tag does not appear as orphaned in the tree (as a dataset does) until the tree is refreshed - see screenshots 1 and 2

-- Dataset / Project

Can cut a dataset inside a project and then paste it back into another project

-- Image / Dataset

Cannot cut an image when under a tag directly
Can cut an image when in dataset or P/D
After cutting not available for pasting anywhere which is inconsistent with cut-paste of tags

Bug detailed below occurred at this stage and after 1.5 hours of testing I decided to stop at this point.

-- Plate / Screen

Functionality

- Drag and Drop all data types that can be Cut as above

Not tested

- Tagged objects can also be Copied from below Tag (NOT Cut, as in Insight)

Not tested.

- Delete Tag & Tag set

Behaves as expected.

Other issues:

- In the toolbar the Create Tagset icon should be to the left of the Create Tag icon to keep it consistent with P/D/S and with the order in the right-click menu.
- the grey expansion arrows are visible on empty tagset and tags in the tree until they are selected - then it disappears. This is inconsistent with the data tree where empty P/D/S do not have the arrows visible.

Bug:

- after trying to add some test tags and attach them in the data tree got beach ball for about a minute in the tag dialog box. Completed tagging and went to tag tree expanded tag and tried to right-click on image in tree - beachball for 5 minutes - selected console when came back and it showed a number of errors. Went back to data tree and console showed 3 more errors - all looked the same but different from the previous one.
Stopped testing at this point.
Cannot find way of saving error output as text, but have kept window if you need further details.

Jquery errors did not copy so screenshot immediately below this to show these.

Error screenshots 1-3

![webclient 3](https://cloud.githubusercontent.com/assets/1552703/13701501/3dafb202-e780-11e5-917b-a8bd8a3c6d8e.jpg)

![webclient 4](https://cloud.githubusercontent.com/assets/1552703/13701502/3dcac5ba-e780-11e5-8922-c2e398089584.jpg)

![webclient 5](https://cloud.githubusercontent.com/assets/1552703/13701503/3dd00c96-e780-11e5-8f6b-b627bd6f6bea.jpg)

Screenshot 1

![webclient](https://cloud.githubusercontent.com/assets/1552703/13701510/48863c3c-e780-11e5-9e75-58dfa415a5e0.jpg)

Screenshot 2

![webclient 2](https://cloud.githubusercontent.com/assets/1552703/13701516/4d090faa-e780-11e5-96ca-dc5abc13fd82.jpg)

pwalczysko commented:
> Delete Tag & Tag set
Worked as expected.

will-moore commented:
@pwalczysko That last commit fixes Copy toolbar button as mentioned in https://github.com/openmicroscopy/openmicroscopy/pull/4518#issuecomment-195333897

will-moore commented:
@gusferguson Re:

> Minor point: if a tag within a tagset is selected and Create Tag performed I expected the new tag to be created in the same tagset, not orphaned as it does at present. Not a major issue, but I think this would be expected behaviour e.g. from behaviour in Finder/Explorer where new object is created in current location in the hierarchy.

This is the same behaviour as Insight for Tags and Datasets. Could be changed later if all agreed.

will-moore commented:
Previous commit fixes @gusferguson comment:

> All work except when cutting link with tag in tagset - the cut tag does not appear as orphaned in the tree (as a dataset does) until the tree is refreshed - see screenshots 1 and 2

jburel commented:
https://github.com/openmicroscopy/openmicroscopy/pull/4518#issuecomment-195819308
It will better to "get it right" in web in this PR (cf. @gusferguson  comment) and fix it insight.

will-moore commented:
@gusferguson Toolbar Tagset and Tag buttons now switched.
Re:

> the grey expansion arrows are visible on empty tagset and tags in the tree until they are selected - then it disappears. This is inconsistent with the data tree where empty P/D/S do not have the arrows visible.

Currently we don't query the number of tagged objects under each tag in the tree (don't think we ever have in webclient). This would need quite a bit more work on the queries, which should be looked into in another PR.

will-moore commented:
@gusferguson Re:

> After cutting [image] not available for pasting anywhere which is inconsistent with cut-paste of tags

Did you try selecting a Dataset to paste this? This is working for me now.

gusferguson commented:
@will-moore 

Tested using https://cowfish.openmicroscopy.org/merge/webclient/usertags/ user-3 Safari Mac

Tested commits as listed above and repeated the adding tags and activity that gave me spinning beach balls last test, but did not get a repeat of that issue. 

I little bit of a performance hitch - takes up to 20 seconds to load thumbnails of 7 images under one tag - from home on medium connection.
May be worth checking on a faster connection or doing some stress testing.

- Tagset-Tag buttons in expected order in toolbar.
- Copy link button and right-click>edit> copy disabled on orphaned tags.
- Cut and paste still works as expected both buttons and right-click menu.
- Able to select dataset and paste cut image into it.

Up to you whether you want to merge this and do any other bug fixes in a new PR.

pwalczysko commented:
The problem https://github.com/openmicroscopy/openmicroscopy/pull/4518#issuecomment-195333897 - discrepancy between icons and right-click menu is solved now. The Cut icon only is enabled now as expected. Both right-click and the icon work.

will-moore commented:
@gusferguson Now if you create a Tag (when Tag selected) or Dataset (when Dataset selected) the new object will be a sibling of the selected one.

will-moore commented:
@pwalczysko Re: 

> Suggestion: Make the Tagset icon more distinctive from Tag icon (atm they look very similar, hard to see on first look which is which)

I've tweaked the tagset icon to make it slightly more different from tag icon, but I didn't want to make a massive change to ruin the icon.

gusferguson commented:
@pwalczysko @will-moore - do you think having a different colour - or just a different tone of grey would help?

pwalczysko commented:
@gusferguson Yes, it would help.

gusferguson commented:
See issue for discussion on tagset icon options: https://github.com/openmicroscopy/design/issues/26

will-moore commented:
I removed failing Robot test for "Test Copy Paste Orphaned Image" since this is not supported any more, in line with Insight (but against users' expectations?). 
https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/248/robot/Web%20&%20Web/Web_1/Post%20Test/Test%20Copy%20Paste%20Orphaned%20Image/

cc @jburel There has been a lot of discussion on Copy vv Cut, and also whether to disable any P/D/I Cut/Copy/Paste when in the Tag tree, which might make everything a bit clearer. To be discussed...

jburel commented:
Copy vs Cut is certainly something we need to discuss/tackle
do we try to discuss that Thursday post model?

will-moore commented:
As discussed in today's meeting, we now disable copying cutting & dragging any non-Tag objects in the Tag tree. To test:
 - In the Tag tree, shouldn't be able to Cut, Copy or Drag any object except Tag.
 - Check that Edit menu, toolbar Cut/Copy and Dragging is disabled for all non-Tag objects.

gusferguson commented:
@will-moore cc @pwalczysko 

Tested with https://cowfish.openmicroscopy.org/merge/webclient/usertags/ Mac Safari user-3 - own and other user's tags and data in read-annotate-1

Behaves as expected.
Lag times for loading etc. seem a bit better today (connecting from home again).

Ready to merge from my POV - but decision for Will when ready.

chris-allan commented on components/tools/OmeroWeb/omeroweb/webclient/views.py:
```diff
@@ -929,6 +939,10 @@ def create_link(parent_type, parent_id, child_type, child_id):
                     # return remaining links in same format as json above
                     # e.g. {"dataset":{"10":{"image":[1,2,3]}}}
                     for rl in remainingLinks:
+                        if not rl.loaded:
```
Is this still a potential issue? If so it should be resolved, otherwise the proper logging should be put in place or an error condition raised.

will-moore replied:
Sorry, yes, I forgot to clean this up. I'll remove this now...

chris-allan commented:
We should take this opportunity to formally establish the contract of the following methods modified by this PR by integration test:

 * `get_object_links()`
 * `create_link()`
 * `batch_annotate()`
 * `manage_action_containers()`

It would have been ideal to do this before any modifications were made to ensure there are no subtle bugs introduced by the changes to the various conditionals but we will have to do so retroactively now. The only method here that has any coverage I can see at the moment is `manage_action_containers()` and that's a superficial CSRF test.

pwalczysko commented:
Problems detected in IE 11 on Win 7 laptop. 

  - have couple of Tagsets (user-4 on cowfish connected to eel merge)
  - under one of the Tagsets, choose a Tag, e.g. ``5.0 data`` Tag
  - copy it and paste it under another Tagset 
  - in the target tagset, expand the Tag
  - select the first project, observe 404, ``Url: /merge/webclient/metadata_details/project/21917/
We are sorry, but that Project (id:21917) does not exist, or if it does, you have no permission to see it.``
  - when the ID in the message is studied, it turns out that it does not belong to a project, but to an image
  - the bug goes away when the tree is refreshed

will-moore commented:
@chris-allan OK, I think I will start to work on those tests in another PR. That way I'll be writing and testing them locally without this PR in, and they will be tested against this PR by ci.
The first 5.3.0 task after moving existing webclient api/ methods to webgateway is to break up manage_action_containers() into smaller functions linked to new api urls. But having tests for existing functionality will also help to catch bugs in that process.
The ```api_links``` url which uses ```get_object_links()``` and ```create_link()``` is the only part of the existing /api functionality that is missing tests, so they were on my TODO list anyway.

chris-allan commented:
Sounds great.

will-moore commented:
@chris-allan Actually, just started looking at the tests and the ```api_links``` method is actually covered quite well by ```test_links.py``` and CSRF tests of ```manage_action_containers``` are actually testing that various actions work, so we're not in too bad a state. I'll review more and see what's missing...

chris-allan commented:
Yep, seeing that now that I peel back some of the layers. Just for my own sanity I might integration test the underlying methods themselves to avoid having to pipe every condition through the myriad of options and 100+ lines of code that are in `api_links()` but that's up to you.

I accept that `manage_action_containers()` is 300+ line absolute beast. Looking at the changes more closely it would probably suffice to constrain what we are testing for the moment to the `tag` vs. `tagset` differences for `o_type` and the `addnewcontainer` branch of the `if action...` conditional. Making a separate comment on the logic changes there.

Since `create_link()` is currently an inner function it might make sense to pull it out. It's probably self contained enough that it can even be unit tested instead.

There's no coverage at all for `batch_annotate()`, correct?

chris-allan commented on components/tools/OmeroWeb/omeroweb/webclient/views.py:
```diff
@@ -2616,13 +2636,13 @@ def manage_action_containers(request, action, o_type=None, o_id=None,
                                         args=["edit", o_type, o_id]))
-        if o_type is not None and hasattr(manager, o_type) and o_id > 0:
-            # E.g. Parent o_type is 'project'...
+        if o_type == "project" and hasattr(manager, o_type) and o_id > 0:
```
Since the only object we could be creating in the tree at the moment that has the option of also having a parent is a `Dataset` that is the reason for constraining this, correct?

will-moore replied:
Yes, since we don't allow creation of Plates in the tree, and creation of Tag under Tagset is handled below.

chris-allan replied:
Great. Makes sense.

will-moore commented:
@chris-allan I split relevant tests out of test_csrf.py, added a bunch of tests for tags, split ```create_link()``` out of ```api_links()``` and split ```api_links()``` into smaller functions.

The ```batch_annotate()``` is not covered by any integration tests (it just gets HTML for display in right panel) but it is tested by Robot framework.

will-moore commented:
@pwalczysko That last commit should fix the copy/paste 404 bug https://github.com/openmicroscopy/openmicroscopy/pull/4518#issuecomment-198308731.

gusferguson commented:
@will-moore  cc @pwalczysko 

Tested using https://cowfish.openmicroscopy.org/merge/webclient/ user-4

Repeated workflow that gave Petr 404 error on Windows 8.1:

- with IE 11, Firefox and Chrome - behaved as expected - no error.

Repeated on Mac 10.11.3:

- Chrome and Firefox - behaved as expected - no error.
Safari and received same 404 errors as Petr did before - see screenshots.
Resolved with refresh as before.

![webclient](https://cloud.githubusercontent.com/assets/1552703/14041189/709ae2b8-f266-11e5-8fa6-689f439b9e03.png)

![webclient](https://cloud.githubusercontent.com/assets/1552703/14041190/70b6de28-f266-11e5-8292-7d2826b08314.png)

gusferguson commented:
@will-moore @pwalczysko 

Interestingly - I went back to Safari to test the next PR and did a browser reload of the web client.
I could not then replicate the error in Safari - so I suspect this may have been some sort of caching issue.

So I would say this behaves as expected and is good to merge all else being ready.

will-moore commented:
Thanks @gusferguson.
If @chris-allan is happy with the test coverage, then would be good to merge so I can fix conflicts with https://github.com/openmicroscopy/openmicroscopy/pull/4534

will-moore commented:
I'd really like to get this merged so I can work on the Copy / Cut ticket in a new PR. cc @jburel

chris-allan commented:
Happy for now until we fully refactor `api_links()` to aid testability and and add coverage to as much of the result as possible.

jburel commented:
Thanks all. Merging
